File: stream-client/stream-client-python/streamclient.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

from config import Config
from serviceconnector import ServiceConnector, ConnectionErrorChecker
from streamwriter import StreamWriter
import json

class StreamClient(ConnectionErrorChecker):
    __serviceConnector = None
    __serviceConfig = None

    __BASE_VER = '/v2'
    __REQUEST_PLACEHOLDERS = {
        'streamid': '<streamid>'
    }
    __REQUESTS = { 'streams': __BASE_VER + '/streams' }
    __REQUESTS['stream'] = __REQUESTS['streams'] + '/' + __REQUEST_PLACEHOLDERS['streamid']
    __REQUESTS['consumerid'] = __REQUESTS['stream'] + '/consumer-id'
    __REQUESTS['dequeue'] = __REQUESTS['stream'] + '/dequeue'
    __REQUESTS['config'] = __REQUESTS['stream'] + '/config'
    __REQUESTS['info'] = __REQUESTS['stream'] + '/info'
    __REQUESTS['truncate'] = __REQUESTS['stream'] + '/truncate'

    def __init__(self, config = Config()):
        self.__serviceConfig = config
        self.__serviceConnector = ServiceConnector(self.__serviceConfig)

    def __prepareUri(self, requestName, placeholderName = 'streamid', data = ''):
        return self.__REQUESTS[requestName].replace(self.__REQUEST_PLACEHOLDERS[placeholderName], data)

    def create(self, stream):
        """
        Creates a stream with the given name.

        Keyword arguments:
        stream -- stream name to create
        """
        uri = self.__prepareUri('stream', data=stream)

        self.__serviceConnector.request('PUT', uri)

    def setTTL(self, stream, ttl):
        """
        Set the Time-To-Live (TTL) property of the given stream.

        Keyword arguments:
        stream -- stream name to create or to retrieve
        ttl -- Time-To-Live in seconds
        """
        objectToSend = {
            'ttl': ttl
        }
        uri = self.__prepareUri('config', data=stream)
        data = json.dumps(objectToSend)

        self.checkResponseErrors(
            self.__serviceConnector.request('PUT', uri, data)
        )

    def getTTL(self, stream):
        """
        Retrieves the Time-To-Live (TTL) property of the given stream.

        Keyword arguments:
        stream -- stream name to retrieve ttl for

        Return value:
        Time-To-Live property in seconds
        """
        uri = self.__prepareUri('info', data=stream)
        response = self.checkResponseErrors(
            self.__serviceConnector.request('GET', uri)
        )

        ttl = response.json()['ttl']

        return ttl

    def truncate(self, stream):
        """
        Truncates all existing events in the give stream.

        Keyword arguments:
        stream -- stream name to truncate
        """
        uri = self.__prepareUri('truncate', data=stream)

        self.checkResponseErrors(
            self.__serviceConnector.request('POST', uri)
        )

    def createWriter(self, stream):
        """
        Creates a {@link StreamWriter} instance for writing events to the given stream.

        Keyword arguments:
        stream -- stream name to get StreamWrite instance for
        """

        """
        A bit ugly, but effective method to check if stream exists.
        The main idea is there is could not be presented info for
        invalid stream.
        """
        self.getTTL(stream)

        uri = self.__prepareUri('stream', data=stream)

        return StreamWriter(
            ServiceConnector(self.__serviceConfig),
            uri
        )

File: stream-client/stream-client-ruby/lib/cdap-stream-client/stream_client.rb
require 'cdap-stream-client/stream_writer'

module CDAP

  ###
  # The client interface to interact with services provided by stream endpoint.
  class StreamClient

    attr_reader :rest

    def initialize config = {}
      @rest = Rest.new config
    end

    ###
    # Creates a stream with the given name.
    def create(stream)
      rest.request 'put', stream
    end

    ###
    # Set the Time-To-Live (TTL) property of the given stream.
    #
    # @param stream Name of the stream
    # @param ttl TTL in seconds
    # @throws NotFoundException If the stream does not exists
    def set_ttl(stream, ttl)
      rest.request 'put', "#{stream}/config", body: { ttl: ttl }
    end

    ###
    # Retrieves the Time-To-Live (TTL) property of the given stream.
    #
    # @param stream Name of the stream
    # @return Current TTL of the stream in seconds
    # @throws NotFoundException If the stream does not exists
    def get_ttl(stream)
      response = rest.request 'get', "#{stream}/info"
      response['ttl']
    end

    ###
    # Truncates all existing events in the give stream.
    #
    # @param stream Name of the stream
    # @throws NotFoundException If the stream does not exists
    def truncate(stream)
      rest.request 'post', "#{stream}/truncate"
    end

    ###
    # Creates a {@link StreamWriter} instance for writing events to the given stream.
    #
    # @param stream Name of the stream
    # @return An instance of {@link StreamWriter} that is ready for writing events to the stream
    # @throws NotFoundException If the stream does not exists
    def create_writer(stream, pool_size = nil)
      StreamWriter.new stream, rest, pool_size
    end

  end

end
